Plugins/Creature/Extras/BonusStacking.cpp

```diff
@@ -0,0 +1,569 @@
+#include "Extras/BonusStacking.hpp"
+#include "Services/Hooks/Hooks.hpp"
+#include "Utils.hpp"
+#include "API/CAppManager.hpp"
+#include "API/CNWSAreaOfEffectObject.hpp"
+#include "API/CNWSCombatRound.hpp"
+#include "API/CNWSCreature.hpp"
+#include "API/CNWSCreatureStats.hpp"
+#include "API/CNWSItem.hpp"
+#include "API/CNWSObject.hpp"
+#include "API/CNWSpellArray.hpp"
+#include "API/CNWSRules.hpp"
+#include "API/CExoArrayList.hpp"
+#include "API/CServerExoApp.hpp"
+
+namespace Creature
+{
+
+using namespace NWNXLib;
+using namespace NWNXLib::API;
+using ArgumentStack = NWNXLib::Services::Events::ArgumentStack;
+
+int BonusStacking::s_nAbilityStackingMode = NostackMode::DISABLED;
+int BonusStacking::s_nSkillStackingMode = NostackMode::DISABLED;
+int BonusStacking::s_nSavingThrowStackingMode = NostackMode::DISABLED;
+int BonusStacking::s_nAttackBonusStackingMode = NostackMode::DISABLED;
+bool BonusStacking::s_bAlwaysStackPenalties = false;
+std::vector<int> g_nSpellBonusTypes;
+NostackType::TYPE g_nDefaultType = NostackType::ENHANCEMENT;
+
+NWNXLib::Hooking::FunctionHook* pGetTotalEffectBonusHook;
+
+void BonusStacking::Init(Services::HooksProxy* hooker, int nAbilityStacking, int nSkillStacking, int nSavingThrowStacking, int nAttackBonusStacking, int nDefaultType, bool bAlwaysStackPenalties)
+{
+    s_nAbilityStackingMode = nAbilityStacking;
+    s_nSkillStackingMode = nSkillStacking;
+    s_nSavingThrowStackingMode = nSavingThrowStacking;
+    s_nAttackBonusStackingMode = nAttackBonusStacking;
+    s_bAlwaysStackPenalties = bAlwaysStackPenalties;
+    g_nDefaultType = static_cast<NostackType::TYPE>(nDefaultType);
+
+    LOG_INFO("Property effect stacking modes -- Ability scores: %d | Skill bonuses: %d | Saving throw bonuses: %d"
+        " | Attack bonuses: %d", nAbilityStacking, nSkillStacking, nSavingThrowStacking, nAttackBonusStacking);
+
+    hooker->RequestExclusiveHook<Functions::_ZN12CNWSCreature19GetTotalEffectBonusEhP10CNWSObjectiihhhhi>
+        (&GetTotalEffectBonus);
+    pGetTotalEffectBonusHook = hooker->FindHookByAddress(Functions::_ZN12CNWSCreature19GetTotalEffectBonusEhP10CNWSObjectiihhhhi);
+}
+
+inline __attribute__((always_inline)) bool CheckRaceAlignment(uint16_t nRace, uint16_t nEffectRace, uint8_t nAlignLaw,
+    uint8_t nEffectAlignLaw, uint8_t nAlignGood, uint8_t nEffectAlignGood);
+inline __attribute__((always_inline)) void AddPositiveEffect(uint32_t nSpellId, uint32_t nEffectStrength);
```

Comment:
If you are asking it probably isn't needed. I though that since the function is called frequently, it would be better to avoid the extra stack frames.

I'll leave just the hint and let the compiler figure it out.